### DISCUSSION TITLE
Querying a Slingbox for the current configuration? (video source and remote BIN in particular)

### DISCUSSION THREAD

**Discussion by DVoorhees1957** [GerryDazoo/Slinger#424] [Q&A]
Is there a way to query a Slingbox for its current configuration, particularly the video source (component, HDMI,...) and the current remote BIN that's loaded? Slingplayer Desktop used to be (and may still be) able to provide this info, but I was hoping this info could be found without using Slingplayer Desktop... maybe a REST API request?

EDIT: Maybe there's a way to use rbf121's slinger_utility.py code to display this info?

**Comment by GerryDazoo** [comment]
There is a function in the SDK to do this but we have not yet been able to make it work.

**Reply by DVoorhees1957** [reply]
Slingplayer Desktop must issue the corresponding command(s) somewhere in the packet capture that you already have, yes? Or does it not bother, if it thinks it's doing a new setup?

**Reply by GerryDazoo** [reply]
Yes, probably. Here's a new capture of just the remote setup for my M1. It starts after I've selected the remote type. I can easily get captures of setting the slingbox name too.
[M1_remote_config.zip](https://github.com/GerryDazoo/Slinger/files/10184622/M1_remote_config.zip)

Somehow Desktop 5.0 is working for me. Can you try on your system,

![image](https://user-images.githubusercontent.com/102599361/206428671-5000fc74-fc50-467d-b4ac-146e049be894.png)

**Reply by DVoorhees1957** [reply]
I'm remote from my Slingboxes; it will be several weeks before I'm back on the same LAN as them. I tried to run Slingplayer Desktop; it said it was "Searching for Slingbox devices", then eventually gave the error message "Failed to load Slingplayer Desktop. Error code is -118. Please check your Internet connection and try again."

**Reply by GerryDazoo** [reply]
Yes, you have to be on the LAN where the slingbox is located for that to work.

**Reply by DVoorhees1957** [reply]
Looking at the M1 setup packet capture, I assume the IPs are different than for the 500 setup? It looks like 192.168.117.113 is the device running Slingplayer Desktop, and 192.168.117.115 is the M1...?

EDIT: I'm still trying to figure out the "protocol" for reconfiguring the AV input and/or the remote. From this packet capture, frame 13 shows a GET sent from 192.168.117.113 to 54.213.96.66 (a Sling-controlled IP?), with a bunch of detailed info about Slingplayer Desktop and the target Slingbox:
```
Request URI Query [truncated]: table_name=box_setup_app&client_name=SPD&client_version=5.3.27.297&web_session_id=7590844943188131&user_id=gerry%40dazoo.net&locale=en_CA&plugin_version=2.4.0.133&wifi_plugin_version=1.0.0.8&client_user_agen
    Request URI Query Parameter: table_name=box_setup_app
    Request URI Query Parameter: client_name=SPD
    Request URI Query Parameter: client_version=5.3.27.297
    Request URI Query Parameter: web_session_id=7590844943188131
    Request URI Query Parameter: user_id=gerry%40dazoo.net
    Request URI Query Parameter: locale=en_CA
    Request URI Query Parameter: plugin_version=2.4.0.133
    Request URI Query Parameter: wifi_plugin_version=1.0.0.8
    Request URI Query Parameter: client_user_agent=Mozilla%2F5.0%20(Windows%20NT%206.3%3B%20WOW64)%20AppleWebKit%2F537.36%20(KHTML%2C%20like%20Gecko)%20Chrome%2F39.0.2171.36%20Safari%2F537.36
    Request URI Query Parameter: page_name=VideoSignalFoundTemplate
    Request URI Query Parameter: os_name=WINDOWS
    Request URI Query Parameter: os_version=NT%206.3
    Request URI Query Parameter: os_bits=Win32
    Request URI Query Parameter: browser_version=5.0.0.83
    Request URI Query Parameter: box_fw_version=01.02.092
    Request URI Query Parameter: box_hw_version=01.00.0000
    Request URI Query Parameter: box_user_type=admin
    Request URI Query Parameter: finder_id=E706A7405882C03A076502246DAD632F
    Request URI Query Parameter: box_name=M1
    Request URI Query Parameter: product_id=32
    Request URI Query Parameter: setup_type=avReconfig
    Request URI Query Parameter: unique_machine_id=374D6E82D9994FC9BD72EDCC1298EDFE
    Request URI Query Parameter: event_name=VIDEO_FOUND_CONTINUE_BUTTON_CLICKED
    Request URI Query Parameter: event_start_time=2022-12-08%2010%3A40%3A07
    Request URI Query Parameter: event_end_time=2022-12-08%2010%3A40%3A24
    Request URI Query Parameter: event_status=SUCCESS
    Request URI Query Parameter: event_duration_ms=17329
    Request URI Query Parameter: seq_num=76
    Request URI Query Parameter: rnd=0.30165103171020746
```
The next frame, frame 14, shows a POST from 192.168.117.113 to 192.168.117.115 with the following info:
```
POST /slingbox/15572/setup/avsetup/avinputs/component?forceOkStatus&account=admin&counter=123&cnonce=1670496024885&digest=a2d353ef06ade9de56d2a8f676f7dd71

{"action":"configure","configure_data":{"provider_data":"",
"device_type":"cable_box",
"ir_setup_code":1283,
"brand_name":"Motorola",
"model_name":"DCX3400M",
"number_of_digits":0,
"channel_enter_required":0,
"inter_key_delay":0,
"av_context":"PVR",
"ir_device_guid":""}}
```
which basically makes it look like Slingplayer Desktop is telling the Slingbox that it's going to get remote info for a Motorola DCX3400M (with BIN file name C1283_PL.bin - the "C" comes from device_type cable_box, and most BIN files destined for a Slingbox, at least in the US, have a "_PL.bin" ending to the file name).

Still deciphering what follows these packets... it sorta looks like the BIN file comes from an external IP rather than the Slingplayer Desktop machine. Do you happen to have a capture that attempts to provide a "custom" BIN file that's stored on the same PC/Mac/laptop that's running Slingplayer Desktop? I think that would be particularly relevant for what we're trying to accomplish...

**Reply by GerryDazoo** [reply]
Yes, 113 was running desktop  115 was the M1.
I can do a custom bin capture for you.

**Reply by DVoorhees1957** [reply]
OK, frame 28 has the followup interaction between SPD and the Slingbox - not sure where the desktop got the binary data that it feeds to the Slingbox (probably somewhere between frame 14 and frame 28, lol)... and I'm not sure how to decipher the "remote_bin_utf8" key value pair to make sure it matches what I have as C1283_PL.bin:
```
POST /slingbox/15572/setup/avsetup/avinputs/component?forceOkStatus&account=admin&counter=126&cnonce=1670496026041&digest=612ef288ae86155a3bde71bcf938ad64 HTTP/1.1
Host: 192.168.117.115:5301
Connection: keep-alive
Content-Length: 425
Accept: application/json, text/javascript, */*; q=0.01
Origin: http://apollo.slingbox.com
User-Agent: Mozilla/5.0 (Windows NT 6.3; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/39.0.2171.36 Safari/537.36
Content-Type: application/x-www-form-urlencoded; charset=UTF-8
Referer: http://apollo.slingbox.com/desktop/index?locale=en_CA
Accept-Encoding: gzip, deflate
Accept-Language: en-us,en;q=0.8

{"action":"configure_remote",
"configure_remote_data":
{"device_type":"cable_box",
"ir_setup_code":1283,
"remote_bin_utf8":"IIC2L4pvGy0uj+8ulQ5fNDKA/jZvf2+v7y9wsPAwcbGy8jJ3iPi4w0S39zd4A4Q4ebl7+zt8u7z5BINDxIVFhsUFh8cHP9HL0bBvb8mvSHHxq59w8Z+Hn/JwMJ+RnxyhceJzsUmwqPGxm/GfYCiy7yzuQCCtFmgsr4Usr/EAcY9xoU2qPAY+AHFzb58EnzIEMKuAnzCKLyzuhiCtNGssr4WCcCzuiizuigoB7HbzUq+CLO5IayDg+AQfMkmxIK1v/iyvhfns2g==",
"date_stamp":"",
"data_chunk_no":0}}
```

**Reply by GerryDazoo** [reply]
I'll see what I can do. I've never loaded a custom bin from the local drive only from the web.

**Reply by DVoorhees1957** [reply]
Probably not too important - there has to be a point in the process where the BIN file has landed on the Slingplayer Desktop device, albeit in a temp folder, and then the process from that point is identical - unless the SPD device can somehow tell Mr. Sling to directly send the BIN file to the Slingbox...

**Reply by GerryDazoo** [reply]
[M1_custom_remote_config_2010.zip](https://github.com/GerryDazoo/Slinger/files/10196795/M1_custom_remote_config_2010.zip)

Custom remote C2010_PL

**Comment by GerryDazoo** [comment]
I know the player downloads the remote control image file and the keycode mappings from the slingbox.
